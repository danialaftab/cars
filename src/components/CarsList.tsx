import React from 'react';
import { ICar } from '../interfaces/ICar';
import { Container, Row } from 'react-bootstrap';

const CarsList = (props: any) => {

    let renderCars = () => {
        let { cars } = props;

        cars = cars.map((car: ICar, idx: number) =>
            <div key={idx} className="carCard">
                <div className="img">
                    <img src={car.pictureUrl} alt="car_image" />
                </div>
                <div className="body">
                    <h4>
                        {car.manufacturerName + ' ' + car.modelName}
                    </h4>
                    <p>Stock # {car.stockNumber + ' - ' + car.mileage.number + ' ' + car.mileage.unit + ' - ' + car.fuelType + ' - ' + car.color}</p>
                    <a href={`/detail/${car.stockNumber}`}>
                        View Details
                    </a>
                </div>
            </div>

        );

        return cars
    }

    return (
        <Container>
            <div>
                <h4 className="contentHeading">Available Cars</h4>
                    <p>Showing {props.cars.length} of {props.totalCars} Results </p>
            </div>
            <br />
            <Row>
                {renderCars()}
            </Row>
        </Container>
    )
}

export default CarsList;