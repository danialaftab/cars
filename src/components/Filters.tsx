import React, { useState, useEffect } from 'react';
import { getColors, getManufacturers } from '../Api';
import { Form, Button, Spinner } from 'react-bootstrap';

type UpdateFilterType = (manufacturer: string, color: string) => any

const Filters = (props: { updateFilters : UpdateFilterType}) => {
    const [colors, setColors] = useState([]);
    const [manufacturers, setManufacturers] = useState([]);
    const [selectedColor, setSelectedColor] = useState('');
    const [selectedManufacturer, setSelectedManufacturer] = useState('');
    const [isColorsLoading, setIsColorsLoading] = useState(false);
    const [isManufacturerLoading, setIsManufacturerLoading] = useState(false);


    useEffect(() => {
        async function fetchColors() {
            setIsColorsLoading(true);
            let col = await getColors();
            setIsColorsLoading(false);
            setColors(col.colors);
        }

        async function fetchManufacturers() {
            setIsManufacturerLoading(true);
            let manu = await getManufacturers();
            setIsManufacturerLoading(false);
            manu = manu.manufacturers.map((manufacturer: any) => manufacturer.name)
            setManufacturers(manu)
        }

        fetchColors()
        fetchManufacturers()
    }, [])


    return (
        <div className="filterBox" data-testid="filters">
            <Form>
                <Form.Group>
                    <Form.Label>Color</Form.Label>
                    <br />
                    {isColorsLoading && <Spinner animation="border" />}
                    {!isColorsLoading && <Form.Control as="select" onChange={(e) => setSelectedColor(e.target.value)}>
                        <option value=' '>All car colors</option>
                        {colors.map(color => <option>{color}</option>)}
                    </Form.Control>}

                    <Form.Label>Manufacturer</Form.Label>
                    <br />
                    {isManufacturerLoading && <Spinner animation="border" />}
                    {!isManufacturerLoading && <Form.Control as="select" onChange={(e) => setSelectedManufacturer(e.target.value)}>
                        <option value=' '>All manufacturers</option>
                        {manufacturers.map(manufacturer => <option>{manufacturer}</option>)}
                    </Form.Control>}

                    <br/>   
                    <Button onClick={(e) => { e.preventDefault(); props.updateFilters(selectedManufacturer, selectedColor) }} variant="primary" type="submit">
                        Filter
                </Button>
                </Form.Group>
            </Form>
        </div>
    )
}

export default Filters;