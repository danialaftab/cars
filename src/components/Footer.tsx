import React from 'react';
import { Navbar, Nav } from 'react-bootstrap';

const Footer = () => {

    return (
        <Navbar bg="light" variant="light"  className="justify-content-center">
            <Nav>
                <p className="text-center mt-4 mb-4">Â© Auto 2020</p>
            </Nav>
        </Navbar>
    )
}

export default Footer;