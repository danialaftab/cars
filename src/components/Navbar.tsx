import React from 'react';
import { Navbar, Nav } from 'react-bootstrap';

const SystemNavbar = () => {

    return (
        <Navbar bg="light" variant="light">
            <Navbar.Brand href="/">Auto</Navbar.Brand>
            <Navbar.Collapse className="justify-content-end">
                <Nav>
                    <Nav.Link href="#">Purchase</Nav.Link>
                    <Nav.Link href="#">My Orders</Nav.Link>
                    <Nav.Link href="#">Sell</Nav.Link>
                </Nav>
            </Navbar.Collapse>
        </Navbar>
    )
}

export default SystemNavbar;