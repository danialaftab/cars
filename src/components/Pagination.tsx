import React, { useState, useEffect } from 'react';
import { Pagination } from 'react-bootstrap';

const CarsPagination = (props: any) => {
    const [pages, setPages] = useState<any>([]);
    let { currentPage, totalPages } = props;

    useEffect(() => {
        let pages = [];
        let startingPage = (Math.floor(currentPage / 10) * 10);
        if(startingPage === 0)
            startingPage = 1
    
        for (let i = startingPage; (i < (startingPage + 10)) && (i <= totalPages); i++) {
            pages.push(<Pagination.Item key={i} active={(currentPage === (i))} onClick={() => { props.openPage(i) }}>{i}</Pagination.Item>)
        }

        setPages(pages);
        
    }, [currentPage, totalPages])


    return (
        <Pagination>
            {(currentPage > 10) ? <Pagination.First onClick={() => {props.openPage(1)}} /> : null }
            {(currentPage > 10) ? <Pagination.Prev onClick={() => {props.openPage(currentPage - 10)}} /> : null}

            {pages}

            {(totalPages > 10 && currentPage < totalPages) ? <Pagination.Ellipsis /> : null}
            {(totalPages > 10 && currentPage < totalPages) ? <Pagination.Next onClick={() => {props.openPage(currentPage + 10)}}/> : null}
            {(totalPages > 10 && currentPage < totalPages) ? <Pagination.Last onClick={() => {props.openPage(totalPages)}} /> : null}
        </Pagination>
    )
}

export default CarsPagination;

