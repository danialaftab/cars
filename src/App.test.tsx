import React from 'react';
import { getByTestId, render, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Filters from './components/Filters';
import CarsList from './components/CarsList';
import Footer from "./components/Footer";
import Navbar from "./components/Navbar";
import Pagination from "./components/Pagination";

afterEach(cleanup);

test('Filters match snapshot', () => {
    const tree = renderer.create(<Filters updateFilters={() => {}}></Filters>).toJSON();
    expect(tree).toMatchSnapshot();
});

test('Car list match snapshot', () => {
  const tree = renderer.create(<CarsList cars={[{
    stockNumber: '100w',
    manufacturerName:'tesla',
    modelName:	'model s',
    mileage: {
      number: 1,
      unit: 'km'
    },
    fuelType:	'petrol',
    color:	'white',
    pictureUrl:	'www.example.com'
  }]}></CarsList>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Footer match snapshot', () => {
  const tree = renderer.create(<Footer></Footer>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('NavBar match snapshot', () => {
  const tree = renderer.create(<Navbar></Navbar>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Pagniation match snapshot', () => {
  const tree = renderer.create(<Pagination  currentPage={1}
    totalPages={10}></Pagination>).toJSON();
  expect(tree).toMatchSnapshot();
});



