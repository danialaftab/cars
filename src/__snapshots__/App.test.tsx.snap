// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Car list match snapshot 1`] = `
<div
  className="container"
>
  <div>
    <h4
      className="contentHeading"
    >
      Available Cars
    </h4>
    <p>
      Showing 
      1
       of 
       Results 
    </p>
  </div>
  <br />
  <div
    className="row"
  >
    <div
      className="carCard"
    >
      <div
        className="img"
      >
        <img
          alt="car_image"
          src="www.example.com"
        />
      </div>
      <div
        className="body"
      >
        <h4>
          tesla model s
        </h4>
        <p>
          Stock # 
          100w - 1 km - petrol - white
        </p>
        <a
          href="/detail/100w"
        >
          View Details
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Filters match snapshot 1`] = `
<div
  className="filterBox"
  data-testid="filters"
>
  <form
    className=""
  >
    <div
      className="form-group"
    >
      <label
        className="form-label"
      >
        Color
      </label>
      <br />
      <select
        className="form-control"
        onChange={[Function]}
      >
        <option
          value=" "
        >
          All car colors
        </option>
      </select>
      <label
        className="form-label"
      >
        Manufacturer
      </label>
      <br />
      <select
        className="form-control"
        onChange={[Function]}
      >
        <option
          value=" "
        >
          All manufacturers
        </option>
      </select>
      <br />
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Filter
      </button>
    </div>
  </form>
</div>
`;

exports[`Footer match snapshot 1`] = `
<nav
  className="justify-content-center navbar navbar-expand navbar-light bg-light"
>
  <div
    className="navbar-nav"
    onKeyDown={[Function]}
  >
    <p
      className="text-center mt-4 mb-4"
    >
      Â© Auto 2020
    </p>
  </div>
</nav>
`;

exports[`NavBar match snapshot 1`] = `
<nav
  className="navbar navbar-expand navbar-light bg-light"
>
  <a
    className="navbar-brand"
    href="/"
  >
    Auto
  </a>
  <div
    aria-expanded={null}
    className="justify-content-end navbar-collapse collapse"
  >
    <div
      className="navbar-nav"
      onKeyDown={[Function]}
    >
      <a
        className="nav-link"
        data-rb-event-key="#"
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        Purchase
      </a>
      <a
        className="nav-link"
        data-rb-event-key="#"
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        My Orders
      </a>
      <a
        className="nav-link"
        data-rb-event-key="#"
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        Sell
      </a>
    </div>
  </div>
</nav>
`;

exports[`Pagniation match snapshot 1`] = `
<ul
  className="pagination"
/>
`;
